### ISSUE TITLE
Changing OAuth2 access token settings in one collection or workspace updates other collections and workspaces

### ISSUE THREAD

**Issue by StephenNay** [a85/POSTMan-Chrome-Extension#4636]
Please read through the [guidelines](https://github.com/postmanlabs/postman-app-support#guidelines-for-reporting-issues) before creating a new issue.

**Describe the bug**
I have two collections that use OAuth2 access tokens. The collections interact with different authentication systems at different URLs. When I make changes in the Get New Access Token dialog for one collection, the changes are also made to the other collection.

**To Reproduce**
1. Right-click on a collection and select Edit.
2. Click on 'the Authorization tab. Should be set to OAuth 2.0.
3. Click Get New Access Token
4. Set values such as Access Token URL, Client ID, Client Secret.
5. Click Request Token.
6. Click Use Token.
7. Click Update to close the Edit Collection window.
8. Right-click on a different collection that also uses OAuth 2.0 authentication and select Edit.
9. Click on Authorization tab.
10. Click Get New Access Token.

Result: The Access Token URL, Client ID, etc. in this second collection have been changed like those in the first collection.

On further testing, I discovered that even separate workspaces seem to share the OAuth2 settings!

**Expected behavior**
I expect each collection to have its own authentication settings, and that those settings would not be shared with other collections or updated by changes in other collections.

Even if collections within a workspace have to share settings for some unknown reason, I would absolutely expect different workspaces to have entirely different authentication settings. Why would changes to a non-shared collection in one workspace affect anything at all in the other workspace?? What is the point of having separate workspaces if not to isolate one from the other?

**Screenshots**
none

**App information (please complete the following information):**
 - App Type Native App
 - Postman Version 6.1.3
 - OS: macOS Sierra 10.12.6

**Comment by bogadopedro** [comment]
I have the same situation. I have to rewrite the configuration every time

**Comment by harryi3t** [comment]
This is expected behaviour. The Oauth2 configuration is global to the app.
I am marking this as a feature request.

**Comment by aviweiss** [comment]
I too have made this same request. Each collection ideally would have its own authentication configuration, since many clearly use collections as logical grouping of commands for a given system/api/service etc. 

Additionally, it would be nice to have templatable “sprints” auto-key key names for each collections: so for dev box: “dev -%t” would generate the key name “dev 10:03” and “live - %d %t” would create “live - 2018-07-25 13:14”. This would help in knowing which key is current. Would also be nice to have expired keys “auto delete”. Also, the “use key” does not properly set the key to be used. It must also be selected and visible in key window.

**Comment by mrorat** [comment]
Same here, it would be amazing if we could have more than one config for oAuth

**Comment by aviweiss** [comment]
Yes;

Collections should be able to be based on a API URL, and each API URL should have an oAuth URL used to get a token for that API. 

I’m not quite sure why this modality wasn't envisaged from the beginning, as Postman is touted as a tool for testing APIs, and I have to imagine API developers are frequently testing against more than one API.

**Comment by amex132** [comment]
I'd request any kind of token generation to be completely independent of any request, i.e. not tied to any group etc. I'd like the ability to associate one or more or groups or whatever requests to tokens independently.  
Meaning, it would nice to be able to name a token generation setting and then associate a request with that name so I could associate a request with token generation on the fly.
When executing a request, it should remember the last token generation name.

**Comment by bulat-gab** [comment]
I faced the same problem. Two apps use different authorization urls and I have to rewrite configurations each time I switch the app. Are there any ways to overcome it while we are waiting for the feature?

**Comment by chaodoze** [comment]
+1 each collection should remember Get New Access Token differently

**Comment by lorenzleutgeb** [comment]
Pardon me, but with collections that obviously encourage working on multiple projects it seems quite stupid that authentication cannot be configured the same per-collection/per-project way...

**Comment by masterl1983** [comment]
Any update on this issue? Facing the same problem right now as I have two Azure Tenants where I need to work with and I always have to chnage the oAuth2 Settings which is really not nice :-)

**Comment by chdig** [comment]
This is a drop-dead giant bug that means we cannot use postman.  It is definitely not "expected behavior", or a feature request.  It's simply to do what Postman is advertised to: test multiple APIs.  Since most APIs have their own different authentication/token retrieval, Postman doesn't actually allow seamless use of multiple APIs/collections.

The bugs run deep:
The user experience is very confusing. If I edit a collection, go to the authentication tab, and fill in these settings, they are a) not saved for all users on a team, and b) will appear to overwrite the settings already entered for a different collection/api.

This means that the settings to load a login page have to be entered by each user. Expected behavior would be for settings to be shared, but tokens to be kept separate for users.  Instead it's the opposite: settings are not shared, but for some reason tokens are able to be used for any API, by any user in a postman team.

In the span of 20 minutes, I'd like to be able to test 3 different APIs with their own oauth2 authentication -- it's what Postman is advertised to do.  This basic expectation is not possible without re-setting a full 8 form fields every single time I want to connect to a different API.  Plus the saved tokens are mixed up between different connections.

Could a Postman admin give an update on this issue with when and if it will be resolved?

**Comment by aviweiss** [comment]
Concur.

Postman makes a big deal out of their tool set that in theory should allow for testing different APIs by different people and teams, yet have an oauth implementation that assumes only ONE api via one access token path will be accessed by one person at any given time, which is an extremely silly expectation. 

It’s like buying a Bugatti Veyron and equipping it with 4 temporary spare tires... 

-Avi

**Comment by terramare** [comment]
In a day I too will need to hit 2 or 3 different OAuth 2 services. Doing this in Postman is painful.

It does seem the OAuth config should fit the notion of Workspaces creating isolation within Postman, if not even Collections.

**Comment by coccoinomane** [comment]
Hello,

Please consider allowing the definition of per-collection or per-environment OAuth2 configurations.

Re-write the OAuth2 configuration everytime I change API is time consuming.

Thank you,
Guido

**Comment by nikdea** [comment]
Hi guys, this feature is definitely very userful BUT if you store your OAUTH2 configuration details into environment variables you achieve the same result. ;)

**Comment by aviweiss** [comment]
not sure what env vars you are referring to, or what they have to do with having collection-specific OAUTH2 auth token URLs…

**Comment by coccoinomane** [comment]
Thanks @nicolapietropaolodeagostini, using Environment variables in Postman OAuth Screen worked great 👍🏼

@aviweiss Hopefully this screenshot will make it clearer what Nicola means:

<img width="698" alt="screen shot 2019-02-09 at 16 48 50" src="https://user-images.githubusercontent.com/3920436/52522843-0657bb80-2c8b-11e9-8b08-9f401b7ed259.png">

The orange bits are environment variables, which you can set after you create an environment clicking on the cog icon in the top right corner of Postman window.

This is how my environment variables look like:

<img width="1760" alt="screen shot 2019-02-09 at 16 54 06" src="https://user-images.githubusercontent.com/3920436/52522887-651d3500-2c8b-11e9-9d8c-8b452d3bc744.png">

Cheers,
Guido

**Comment by aviweiss** [comment]
Guido;

Ok, not a bad substitute for having auth associated with collections, which I think is still the more “logical” location for it, given that collections are used as the atomic unit for an API relationship. But environment is readily switched so works for me.

Thanks.

**Comment by nickgsoft** [comment]
What does the Available Tokens dropdown do then? 
https://learning.getpostman.com/docs/postman/sending_api_requests/authorization/
according to this article "In the "Available Tokens" drop-down menu, select an existing token and click the Send button." , but this doesn't switch the token for me. Any idea?

**Comment by aviweiss** [comment]
In my opinion, the authorization panel is in need of a complete overhaul.

In addition to the bug you mention (“use” does not select the token), there are numerous other issues with it:
* inability to have a timestamp variable in the token name to auto label each token with a timestamp
* inability to CLEAR all “expired” tokens with one click
* inability to have an authorization url change as a function of collection selection (not just “environment” selection)
* inability to have authorization automatically run (with appropriate auth url) as initialization of a collection run.

Im not sure if anyone from postman reads these emails, or the bug reports, of which Ive filed numerous amounts of, but still isn’t working well to optimize postman use

**Comment by cevjg** [comment]
i think same as aviweiss.
You can add folder in collection and parameter authorisation for folder with "Inherit from parent" or not.  
But not inherit change configuration from parent. That is a bug!

**Comment by jpsv** [comment]
I'm using Postman to run a series of tasks across different services on an as-needed basis, and this issue affects me.

It's a single collection and all requests are intended to do the same thing to a designated user account in a timely and efficient manner. But because OAuth2.0 is becoming more common, kind of hurts.

**Comment by JanDoggen** [comment]
> 
> 
> What does the Available Tokens dropdown do then?
> https://learning.getpostman.com/docs/postman/sending_api_requests/authorization/
> according to this article "In the "Available Tokens" drop-down menu, select an existing token and click the Send button." , but this doesn't switch the token for me. Any idea?

I have also made Token Name an env var, that makes it easy to select the correct one.

**Comment by tofalck** [comment]
> Thanks @nicolapietropaolodeagostini, using Environment variables in Postman OAuth Screen worked great 👍🏼
> 
> @aviweiss Hopefully this screenshot will make it clearer what Nicola means:
> 
> <img alt="screen shot 2019-02-09 at 16 48 50" width="698" src="https://user-images.githubusercontent.com/3920436/52522843-0657bb80-2c8b-11e9-8b08-9f401b7ed259.png">
> 
> The orange bits are environment variables, which you can set after you create an environment clicking on the cog icon in the top right corner of Postman window.
> 
> This is how my environment variables look like:
> 
> <img alt="screen shot 2019-02-09 at 16 54 06" width="1760" src="https://user-images.githubusercontent.com/3920436/52522887-651d3500-2c8b-11e9-9d8c-8b452d3bc744.png">
> 
> Cheers,
> Guido

The environment work around does not suffice if you work with different Grant Types.

I agree that this seems like something that should be fixed - and it really shouldn't take that long to do, since the functionality is already in there.

This issue has been open now for over 1½ year. Any updates on when it's going to be fixed?

**Comment by aakrem** [comment]
+1

**Comment by rollendxavier** [comment]
The suggested method of "Setup environments" is not what those are meant for. There should be a feature to configure OAuth2 separately for requests.

**Comment by shtrip** [comment]
OAuth authentication has always felt a bit clumsy in Postman. For manual API testing I would love to have a drop down where I can select the OAuth configuration from a predefined set. Similar to selecting an environment.

Having a fixed OAuth configuration per collection would be too limited in my opinion. For collections it should also be easy to switch between OAuth configurations.

**Comment by MarcBKirsch** [comment]
It's a very annoying feature. Even the workaround kind of works for me, inporting collections overrides the parameters potentially. This really needs some love.

**Comment by MarcBKirsch** [comment]
I figured another workaround which am using moving forward: I use a pre-script to generate a token automatically. Took me a few hours to figure out how it works, but I can recommend going down this path. Saves clicks and even better, the requests can be easily be used in collection runs. But I agree, this should be an out of the box feature.

**Comment by jcush** [comment]
For anyone using Grant Type: Password Credentials, I can confirm the env variables method posted by @tofalck does work. Postman doesn't highlight the variable as it does with all the other fields but it does apply it.

For anyone with multiple projects that use _different_ grant types, I can confirm that switching (as I do) between Grant Types (my projects all use either Password Credentials or Client Credentials) _does_ keep the values for the type-specific fields. For example, when you switch type and switch back again your Password variable will still be in place.

@MarcBKirsch I think your method sounds very useful, but it would probably help users here if you post some sort of example (without your credentials of course) to point people in the right direction considering what you've found.

**Comment by MarcBKirsch** [comment]
@jcush - sure, happy to share the pre-req code here that provides the token genration automatically:


`const c_tokenURL = 'https://yourdomain/oauth/token'; // Endpoint that generates the token
const c_clientID = '12345123451234';
const c_clientSecret = '312345678901234';
const c_grantType = 'client_credentials';

// ***************************************************************

var Base64={_keyStr:"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/=",encode:function(e){var t="";var n,r,i,s,o,u,a;var f=0;e=Base64._utf8_encode(e);while(f<e.length){n=e.charCodeAt(f++);r=e.charCodeAt(f++);i=e.charCodeAt(f++);s=n>>2;o=(n&3)<<4|r>>4;u=(r&15)<<2|i>>6;a=i&63;if(isNaN(r)){u=a=64}else if(isNaN(i)){a=64}t=t+this._keyStr.charAt(s)+this._keyStr.charAt(o)+this._keyStr.charAt(u)+this._keyStr.charAt(a)}return t},decode:function(e){var t="";var n,r,i;var s,o,u,a;var f=0;e=e.replace(/[^A-Za-z0-9\+\/\=]/g,"");while(f<e.length){s=this._keyStr.indexOf(e.charAt(f++));o=this._keyStr.indexOf(e.charAt(f++));u=this._keyStr.indexOf(e.charAt(f++));a=this._keyStr.indexOf(e.charAt(f++));n=s<<2|o>>4;r=(o&15)<<4|u>>2;i=(u&3)<<6|a;t=t+String.fromCharCode(n);if(u!=64){t=t+String.fromCharCode(r)}if(a!=64){t=t+String.fromCharCode(i)}}t=Base64._utf8_decode(t);return t},_utf8_encode:function(e){e=e.replace(/\r\n/g,"\n");var t="";for(var n=0;n<e.length;n++){var r=e.charCodeAt(n);if(r<128){t+=String.fromCharCode(r)}else if(r>127&&r<2048){t+=String.fromCharCode(r>>6|192);t+=String.fromCharCode(r&63|128)}else{t+=String.fromCharCode(r>>12|224);t+=String.fromCharCode(r>>6&63|128);t+=String.fromCharCode(r&63|128)}}return t},_utf8_decode:function(e){var t="";var n=0;var r=c1=c2=0;while(n<e.length){r=e.charCodeAt(n);if(r<128){t+=String.fromCharCode(r);n++}else if(r>191&&r<224){c2=e.charCodeAt(n+1);t+=String.fromCharCode((r&31)<<6|c2&63);n+=2}else{c2=e.charCodeAt(n+1);c3=e.charCodeAt(n+2);t+=String.fromCharCode((r&15)<<12|(c2&63)<<6|c3&63);n+=3}}return t}};
var userPass = c_clientID + ':' + c_clientSecret

// Automatic script that will generate a new Toekn at runtime
pm.sendRequest({
      url: c_tokenURL, 
      method: 'POST',
      header: {
        'Accept': 'application/json',
        'Content-Type': 'application/x-www-form-urlencoded',
        'Authorization' : 'Basic ' + Base64.encode(userPass),
      },
      body: {
          mode: 'urlencoded',
          urlencoded: [
            {key: "grant_type", value: c_grantType, disabled: false},

        ]
      }
  }, function (err, res) {
        // Update the token
        pm.environment.set("OAuth_Token", res.json().access_token); //'OAuth_Token' is the environment variable name
  }); `



Make sure you define an environment variable "OAuth_Token" before, or alternatively use a global variable.
The final step is to use the {{OAuth_Token}} in the Token field itself.

![Postman_Example](https://user-images.githubusercontent.com/15914824/79242968-4c06aa00-7e75-11ea-8429-56a67983ebe4.jpg)


I was going down the path of coding the token generation since I have multiple APIs who all have different tokens as well as the need to test against several sandboxes in parallel. My only solution was using this code based approach.

**Comment by ursusursus** [comment]
This has bitten me just now, very illogical to assume 1 global oauth

**Comment by MarcBKirsch** [comment]
@ursusursus - The point is that the OAuth_Token variable is generated at the start of each request. 
It's actually a temporary variable only.
The credentials are stored in this example in each pre-req section. 
Works fine for my 47 requests that I regularly execute in one Collection Run.

Having one global variable is necessary since Postman is supporting only ONE setup of parameters.
If you have a better solution, please share. I might miss something here.

**Comment by ursusursus** [comment]
@MarcBKirsch I wasn't responding to your post sorry

**Comment by TheTallMan67** [comment]
BUMP. Need this functionality

**Comment by mblakesley** [comment]
I also need this!

As a workaround I'm adding separate token endpoints with a post-request script (in the "tests" tab) to save the token received in the response as an environment var. :D

**Comment by nielszz** [comment]
This missing 'feature' is my greatest annoyance with postman! Also sometimes I use the environment variables to migitate this, but this scales poorly when dealing with a lot of different request and OAuth2 endpoints..

**Comment by peter-umb** [comment]
I did not realize this was happening behind the scenes in Postman and I have lost hours worth of work because of this. I can no longer use this as a solution for my work. The other members on my team are now forced to move as well.

**Comment by MarcBKirsch** [comment]
@peter-umb - Have a look above, I shared how the automation of the login flow works. My automation works fantastic for several environments, endpoints and different users. I'm managing a complex environment without any issues.

**Comment by gsacavdm** [comment]
+1 on needing this functionality.

**Comment by surfgirl** [comment]
> @jcush - sure, happy to share the pre-req code here that provides the token genration automatically:
> 
> `const c_tokenURL = 'https://yourdomain/oauth/token'; // Endpoint that generates the token
> const c_clientID = '12345123451234';
> const c_clientSecret = '312345678901234';
> const c_grantType = 'client_credentials';
> 
> // ***************************************************************
> 
> var Base64={_keyStr:"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/=",encode:function(e){var t="";var n,r,i,s,o,u,a;var f=0;e=Base64._utf8_encode(e);while(f<e.length){n=e.charCodeAt(f++);r=e.charCodeAt(f++);i=e.charCodeAt(f++);s=n>>2;o=(n&3)<<4|r>>4;u=(r&15)<<2|i>>6;a=i&63;if(isNaN(r)){u=a=64}else if(isNaN(i)){a=64}t=t+this._keyStr.charAt(s)+this._keyStr.charAt(o)+this._keyStr.charAt(u)+this._keyStr.charAt(a)}return t},decode:function(e){var t="";var n,r,i;var s,o,u,a;var f=0;e=e.replace(/[^A-Za-z0-9+/=]/g,"");while(f<e.length){s=this._keyStr.indexOf(e.charAt(f++));o=this._keyStr.indexOf(e.charAt(f++));u=this._keyStr.indexOf(e.charAt(f++));a=this._keyStr.indexOf(e.charAt(f++));n=s<<2|o>>4;r=(o&15)<<4|u>>2;i=(u&3)<<6|a;t=t+String.fromCharCode(n);if(u!=64){t=t+String.fromCharCode(r)}if(a!=64){t=t+String.fromCharCode(i)}}t=Base64._utf8_decode(t);return t},_utf8_encode:function(e){e=e.replace(/\r\n/g,"\n");var t="";for(var n=0;n<e.length;n++){var r=e.charCodeAt(n);if(r<128){t+=String.fromCharCode(r)}else if(r>127&&r<2048){t+=String.fromCharCode(r>>6|192);t+=String.fromCharCode(r&63|128)}else{t+=String.fromCharCode(r>>12|224);t+=String.fromCharCode(r>>6&63|128);t+=String.fromCharCode(r&63|128)}}return t},_utf8_decode:function(e){var t="";var n=0;var r=c1=c2=0;while(n<e.length){r=e.charCodeAt(n);if(r<128){t+=String.fromCharCode(r);n++}else if(r>191&&r<224){c2=e.charCodeAt(n+1);t+=String.fromCharCode((r&31)<<6|c2&63);n+=2}else{c2=e.charCodeAt(n+1);c3=e.charCodeAt(n+2);t+=String.fromCharCode((r&15)<<12|(c2&63)<<6|c3&63);n+=3}}return t}};
> var userPass = c_clientID + ':' + c_clientSecret
> 
> // Automatic script that will generate a new Toekn at runtime
> pm.sendRequest({
> url: c_tokenURL,
> method: 'POST',
> header: {
> 'Accept': 'application/json',
> 'Content-Type': 'application/x-www-form-urlencoded',
> 'Authorization' : 'Basic ' + Base64.encode(userPass),
> },
> body: {
> mode: 'urlencoded',
> urlencoded: [
> {key: "grant_type", value: c_grantType, disabled: false},
> 
> ```
>     ]
>   }
> ```
> 
> }, function (err, res) {
> // Update the token
> pm.environment.set("OAuth_Token", res.json().access_token); //'OAuth_Token' is the environment variable name
> }); `
> 
> Make sure you define an environment variable "OAuth_Token" before, or alternatively use a global variable.
> The final step is to use the {{OAuth_Token}} in the Token field itself.
> 
> ![Postman_Example](https://user-images.githubusercontent.com/15914824/79242968-4c06aa00-7e75-11ea-8429-56a67983ebe4.jpg)
> 
> I was going down the path of coding the token generation since I have multiple APIs who all have different tokens as well as the need to test against several sandboxes in parallel. My only solution was using this code based approach.

**Comment by surfgirl** [comment]
Hi, any idea, how could I implement that in case of implicit flow? TIA

**Comment by peter-umb** [comment]
Thanks! Maybe I overlooked your post. I will check it out.

**Comment by va-vessels** [comment]
+1 for unique per collection; even better unique per collection and folder unless inherit is selected.

Thanks!

**Comment by vvs11** [comment]
The fix for this is in progress. Will share an update on this soon

**Comment by dcdraper** [comment]
Please, please change this to be unique per folder/request. We have collections where multiple creds are used for different areas of the api based on security roles.

**Comment by vikiCoder** [comment]
Hi everyone,

We have made a few changes in OAuth 2.0 helper in the Postman app to address the problems discussed here. These changes are available in the [Postman Canary](https://www.postman.com/downloads/canary/). Please download the latest Postman Canary app and check it out.

See the issue #9056 to know more about the changes. Feel free to leave your feedback at #9056.

**Comment by vvs11** [comment]
@dcdraper @va-vessels @peter-umb @surfgirl @gsacavdm @nielszz @mblakesley @TheTallMan67 @MarcBKirsch @ursusursus @shtrip @tofalck @aviweiss @jpsv @rollendxavier @cevjg @aviweiss @chdig @masterl1983 @nickgsoft @coccoinomane @mrorat @bogadopedro @StephenNay 
Could you guys refer to the comment by @vikiCoder and try out our canary build and share your feedback? It will really help us figuring out the correct solution to this problem

**Comment by pavlee** [comment]
@vvs11 @vikiCoder
I installed new Canary app, and now it is indeed possible to add OAuth 2.0 configuration to each separate Postman collection. Unrelated: As an additional feature in the future I would like to see possibility of checking if a token is expired and if it is, authenticating for a new token automatically after Send button is clicked on a request that has Authentication configuration.
Right now it is a bit frustrating fetching a new token every now and then because it takes quite a few clicks.

**Comment by vvs11** [comment]
@pavlee Yes, this is something we have in our roadmap. We are figuring out the most effective way of solving this

**Comment by Jonatan-V** [comment]
as va-vessels wrote: +1 for unique per collection; even better unique per collection and folder unless inherit is selected.

**Comment by jthake** [comment]
I really need to be able to have this setting fixed per folder auth setting. So that there is an easy onboarding for a new user of the collection. So if they click "Application" in the window for a request in the Auth tab and it opens the modal screen with the "Get New Access Token", the correct token flow is configured in the next pop up e.g. Client Credentials or Authorization Code.

I dont' find the "Get New Access Token" that intuitive. I would expect to see the config in the first modal to be in the standard auth tab not showing all the config once you click that button.

**Comment by vvs11** [comment]
@jthake-msft So, you'd expect the user to land in the Auth section, select OAuth 2,0, and add all the information other than client credentials and Auth code in the request itself and then click a button to get another popup where the user could enter the credentials and auth code. After this, they can click on the Send button to send the request?
Just want to make sure that I am understanding this correctly.

**Comment by dominicfarr** [comment]
I ran into this one today.  😞 with the rudeness from some in this thread. 

+1 for this feature

**Comment by giridharvc7** [comment]
Closing this issue as this is fixed, tracking OAuth token refresh and expiry in this issue instead #2452

**Comment by giridharvc7** [comment]
@jthake-msft Have created a new ticket for your feature request here https://github.com/postmanlabs/postman-app-support/issues/10575
